libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/introspection/meta_introspection_reader.kt

```diff
@@ -0,0 +1,156 @@
+   */
+  FieldArgsIncludeDeprecated,
+
+  /**
+   * `__InputValue.isDeprecated`
+   */
+  InputValueIsDeprecated,
+
+  /**
+   * `__InputValue.deprecationReason`
+   */
+  InputValueDeprecatedReason,
+
+  /**
+   * `__Type.isOneOf`, introduced in [OneOf Input Objects RFC](https://github.com/graphql/graphql-spec/pull/825/)
+   */
+  TypeIsOneOf
+}
+
+@ApolloInternal
+fun String.toIntrospectionCapabilities(): Set<IntrospectionCapability> {
+  val introspection = this.toMIntrospection()
+  return buildSet {
+    val types = introspection.data.__schema.types
+    types.firstOrNull { it.name == "__Schema" }?.let { schema ->
+      if (schema.fields.orEmpty().any { it.name == "description" }) {
+        add(SchemaDescription)
+      }
+    }
+    types.firstOrNull { it.name == "__Type" }?.let { type ->
+      if (type.fields.orEmpty().any { it.name == "specifiedByURL" }) {
+        add(TypeSpecifiedByURL)
+      }
+      if (type.fields.orEmpty().any { it.name == "isOneOf" }) {
+        add(TypeIsOneOf)
+      }
+      type.fields.orEmpty().firstOrNull{ it.name == "inputFields" }?.let { inputFields ->
+        if (inputFields.args.any { it.name == "includeDeprecated" }) {
+          add(TypeInputFieldsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__Directive" }?.let { directive ->
+      if (directive.fields.orEmpty().any { it.name == "isRepeatable" }) {
+        add(DirectiveIsRepeatable)
+      }
+      directive.fields.orEmpty().firstOrNull{ it.name == "args" }?.let { args ->
+        if (args.args.any { it.name == "includeDeprecated" }) {
+          add(DirectiveArgsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__Field" }?.let { field ->
+      field.fields.orEmpty().firstOrNull{ it.name == "args" }?.let { args ->
+        if (args.args.any { it.name == "includeDeprecated" }) {
+          add(FieldArgsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__InputValue" }?.let { inputValue ->
+      if (inputValue.fields.orEmpty().any { it.name == "isDeprecated" }) {
+        add(InputValueIsDeprecated)
+      }
+      if (inputValue.fields.orEmpty().any { it.name == "deprecationReason" }) {
+        add(InputValueDeprecatedReason)
+      }
+    }
+  }
+}
+
+private val json = Json {
+  // be robust to errors: [] keys
+  ignoreUnknownKeys = true
+}
+
+private fun String.toMIntrospection(): MIntrospection{
+  return json.decodeFromString(this)
+}
+
+@Serializable
+private class MIntrospection(
+    val data: MData
+)
+
+@Serializable
+private class MData(
+    @Suppress("PropertyName")
+    val __schema: MSchema
+)
+
+@Serializable
+private class MSchema(
+    val types: List<MType>,
+)
+
+@Serializable
+private class MType(
+    val name: String,
+    val fields: List<MField>?,
+)
+
+@Serializable
+private class MField(
+    val name: String,
+    val args: List<MInputValue>,
+)
+
+@Serializable
+private class MInputValue(
+    val name: String,
+)
```

Comment:
For the meta introspection query, you could use Apollo and generate all of that automatically.

Reply:
Totally, I was in a "don't use Apollo" mindset, but it makes sense to use it here actually - the code will be simpler 🎉